torchdata/stateful_dataloader/sampler.py

```diff
@@ -89,44 +110,28 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
             assert isinstance(self.sampler_iter, Stateful)
             self.sampler_iter.load_state_dict(state_dict[self._SAMPLER_ITER_STATE])
 
+        if not (isinstance(self.sampler, Stateful) or isinstance(self.sampler_iter, Stateful)) and not isinstance(
+            self.sampler, _InfiniteConstantSampler
+        ):
+            # We skip x samples if underlying sampler is not stateful
+            for _ in range(self.samples_yielded):
+                next(self.sampler_iter)
+
+    def update_state_dict(self) -> None:
+        if isinstance(self.sampler_iter, Stateful) and hasattr(self.sampler_iter, "update_state_dict"):
+            self.sampler_iter.update_state_dict()
```

Comment:
Leaving thoughts here:

* this introduces a new (undocumented) API for samplers (update_state_dict()) which isn't great
* If the alternative solution is to update the base sampler code, this means users bringing their own samplers won't work out of the box, which also isn't great.

Reply:
Thanks!